crates/lang_srv/src/analysis/completion.rs

```diff
@@ -0,0 +1,426 @@
+            debug!(
+                "No specific completionKind for variable type :{:?} defaulting to 'Variable'",
+                a
+            );
+            CompletionItemKind::VARIABLE
+        }
+    };
+
+    CompletionItem {
+        label: str,
+        detail: Some(type_str),
+        kind: Some(typ),
+
+        ..Default::default()
+    }
+}
+///Gets completion items using the visitor pattern,
+///This will walk through declarations that would be accessible from the provided position adding them to a list of completion items until all accessible declarations have been fully explored
+
+pub fn get_completion_items(
+    position: Position,
+    prefix: String,
+    decls: &Declarations,
+    subs: &mut Subs,
+    module_id: &ModuleId,
+    interns: &Interns,
+) -> Vec<CompletionItem> {
+    let completions = get_completions(position, decls, prefix, interns);
+    make_completion_items(
+        subs,
+        module_id,
+        interns,
+        completions
+            .into_iter()
+            .map(|(symb, var)| (symb.as_str(interns).to_string(), var))
+            .collect(),
+    )
+}
+fn make_completion_items(
+    subs: &mut Subs,
+    module_id: &ModuleId,
+    interns: &Interns,
+    completions: Vec<(String, Variable)>,
+) -> Vec<CompletionItem> {
+    completions
+        .into_iter()
+        .map(|(symbol, var)| make_completion_item(subs, module_id, interns, symbol, var))
+        .collect()
+}
+
+///Finds the types of and names of all the fields of a record
+///`var` should be a `Variable` that you know is a record's type or else it will return an empty list
+fn find_record_fields(var: Variable, subs: &mut Subs) -> Vec<(String, Variable)> {
+    let content = subs.get(var);
+    match content.content {
+        roc_types::subs::Content::Structure(typ) => match typ {
+            roc_types::subs::FlatType::Record(fields, ext) => {
+                let field_types = fields.unsorted_iterator(subs, ext);
+
+                match field_types {
+                    Ok(field) => field
+                        .map(|a| (a.0.clone().into(), a.1.into_inner()))
+                        .collect::<Vec<_>>(),
+                    Err(err) => {
+                        warn!("Error getting record field types for completion{:?}", err);
+                        vec![]
+                    }
+                }
+            }
+            roc_types::subs::FlatType::Tuple(elems, ext) => {
+                let elems = elems.unsorted_iterator(subs, ext);
+                let elems = match elems {
+                    Ok(elem) => elem.map(|(num, var)| (num.to_string(), var)).collect(),
+                    Err(err) => {
+                        warn!("Error getting tuple elems for completion{:?}", err);
+                        vec![]
+                    }
+                };
+                elems
+            }
+
+            _ => {
+                warn!("Trying to get field completion for a type that is not a record ",);
```

Comment:
Maybe worth logging the type
```suggestion
                warn!("Trying to get field completion for a type that is not a record : {:?}", subs.dbg(typ));
```